roboteam_observer/observer/src/filters/vision/ball/CameraGroundBallFilter.cpp

```diff
@@ -1,54 +1,134 @@
-    startCovariance(3, 3) = BALL_VELOCITY_INITIAL_COV;
+    startCovariance.diagonal() << INITIAL_POS_COV, INITIAL_POS_COV, INITIAL_VEL_COV, INITIAL_VEL_COV;
 
     constexpr double BALL_MODEL_ERROR = 1.0;
     constexpr double BALL_MEASUREMENT_ERROR = 0.002;  //[m] estimated average position uncertainty in ball detections
     ekf = GroundBallExtendedKalmanFilter(startState, startCovariance, BALL_MODEL_ERROR, BALL_MEASUREMENT_ERROR, observation.timeCaptured);
 }
-CameraGroundBallPrediction CameraGroundBallFilter::predict(Time time) const {
-    const auto &estimate = ekf.getStateEstimate(time);
+
+bool CameraGroundBallFilter::checkRobotCollision(const FilteredRobot& robot, const Eigen::Vector2d& positionEstimate, const Eigen::Vector2d& velocityEstimate) {
+    rtt::Vector2 robotPos = rtt::Vector2(robot.position.position.x(), robot.position.position.y());
+    rtt::Angle robotYaw = rtt::Angle(robot.position.yaw);
+    rtt::Vector2 ballPos = rtt::Vector2(positionEstimate.x(), positionEstimate.y());
+    rtt::Vector2 ballVel = rtt::Vector2(velocityEstimate.x(), velocityEstimate.y());
+    rtt::Vector2 robotVel = rtt::Vector2(robot.velocity.velocity.x(), robot.velocity.velocity.y());
+
+    if (!rtt::RobotShape(robotPos, 0.07, 0.09, robotYaw).contains(ballPos)) {
+        return false;
+    }
+
+    auto ballVelUsed = ballVel - robotVel;
+    auto robotShape = rtt::RobotShape(robotPos, 0.07 + 0.0213, 0.09 + 0.0213, robotYaw);
+    auto frontLine = robotShape.kicker();
+    auto ballVelLine = rtt::LineSegment(ballPos, ballPos - ballVelUsed.stretchToLength(0.09 * 5));
+
+    if (frontLine.intersects(ballVelLine)) {
+        auto collisionAngle = (robotYaw - rtt::Angle(ballVelUsed.scale(-1)));
+
+        if (std::abs(collisionAngle) > 0.5 * M_PI) {
+            return false;
+        }
+
+        auto outVelAbs = ballVelUsed.length() - (ballVelUsed.length() * std::cos(collisionAngle) * 1);
+        auto outVel = rtt::Vector2(robotYaw).scale(-1).rotate(collisionAngle).scale(-1).stretchToLength(outVelAbs);
+        outVel += robotVel;
+        Eigen::Vector2d outVelEigen(outVel.x, outVel.y);
+        ekf.setVelocity(outVelEigen);
+        return true;
+    }
+
+    auto botCircle = rtt::Circle(robotPos, 0.09 + 0.0213);
+    auto intersects = botCircle.intersects(ballVelLine);
+
+    if (!intersects.empty()) {
+        auto intersect = intersects[0];
+        auto collisionAngle = (rtt::Vector2(robotPos - intersect).toAngle() - ballVelUsed.toAngle());
+
+        if (std::abs(collisionAngle) > 0.5 * M_PI) {
+            return false;
+        }
+
+        auto outVelAbs = ballVelUsed.length() - (ballVelUsed.length() * std::cos(collisionAngle) * 0.6);
+        auto outVel = rtt::Vector2(robotPos - intersect).rotate(collisionAngle).scale(-1).stretchToLength(outVelAbs);
+        outVel += robotVel;
+        Eigen::Vector2d outVelEigen(outVel.x, outVel.y);
+        ekf.setVelocity(outVelEigen);
+        return true;
+    }
+
+    return false;
+}
```

Comment:
Maybe want to create a function in utils that tells you if they intersect and if so if they intersect on the kicker or the hull.

Reply:
There is already a function that checks intersection, so this should not be much extra work.